src/libstd/sys/windows/fs.rs

```diff
@@ -567,23 +646,85 @@ pub fn remove_dir_all(path: &Path) -> io::Result<()> {
         // rmdir only deletes dir symlinks and junctions, not file symlinks.
         rmdir(path)
     } else {
-        remove_dir_all_recursive(path)
+        // canonicalize to get a `//?/`-path, which can handle files like `CON`
+        // and `morse .. .`,  and when a directory structure is so deep it needs
+        // long path names
+        let path = try!(path.canonicalize());
+        let base_dir = match path.parent() {
+            Some(dir) => dir,
+            None => return Err(io::Error::new(io::ErrorKind::PermissionDenied,
+                                              "can't delete root directory"))
+        };
+        try!(remove_dir_all_recursive(path.as_ref(), base_dir.as_ref(), 0));
+        Ok(())
     }
 }
 
-fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
+fn remove_dir_all_recursive(path: &Path, base_dir: &Path, mut counter: u64)
+        -> io::Result<u64> {
+    // On Windows it is not enough to just recursively remove the contents of a
+    // directory and then the directory itself. Deleting does not happen
+    // instantaneously, but is scheduled. So just before or after flagging a
+    // file for deletion, we move it to some `base_dir` to avoid races.
+
+    fn move_item(file: &File, base_dir: &Path, mut counter: u64)
+            -> io::Result<u64> {
+        let mut tmpname = base_dir.join(format!{"rm-{}", counter});
+        counter += 1;
+        // Try to rename the file. If it already exists, just retry with an other filename.
+        while let Err(err) = file.rename(tmpname.as_ref(), false) {
+            if err.kind() != io::ErrorKind::AlreadyExists { return Err(err) };
+            tmpname = base_dir.join(format!{"rm-{}", counter});
+            counter += 1;
+        }
+        Ok(counter)
+    }
+
+    fn remove_item(path: &Path,
+                   base_dir: &Path,
+                   mut counter: u64,
+                   readonly: bool) -> io::Result<u64> {
+        if !readonly {
+            let mut opts = OpenOptions::new();
+            opts.access_mode(c::DELETE);
+            opts.custom_flags(c::FILE_FLAG_BACKUP_SEMANTICS | // delete directory
+                              c::FILE_FLAG_OPEN_REPARSE_POINT | // delete symlink
+                              c::FILE_FLAG_DELETE_ON_CLOSE);
+            let file = try!(File::open(path, &opts));
+            counter = try!(move_item(&file, base_dir, counter));
+        } else {
+            let mut opts = OpenOptions::new();
+            opts.access_mode(c::DELETE | c::FILE_WRITE_ATTRIBUTES);
+            opts.custom_flags(c::FILE_FLAG_OPEN_REPARSE_POINT);
+            let file = try!(File::open(path, &opts));
+            // remove read-only permision
+            try!(file.set_perm(FilePermissions::new()));
+            counter = try!(move_item(&file, base_dir, counter));
+            try!(file.remove());
```

Comment:
Could the file be removed before moving? I'm a little worried about moving a file and then the process crashes or something, so these `rm-{tag}` files get littered all over the place